### ISSUE TITLE
problem with variable mrsow in LS3MIP experiments

### ISSUE THREAD

**Issue by francocatalano** [EC-Earth/ece2cmor3#594]
The ece2cmor computation of LS3MIP variable mrsow requires constant field soil type (43.128).
The LS3MIP experiments have been run saving this variable only in the in the +000000 ICMGG output file. For some reason, it looks like ece2cmor3 expects to find field 43.128 in all the output ICMGG files. This gives the following error when runing ece2cmor3:
2020-02-13 11:21:18 ERROR:ece2cmor3.grib_filter: Field missing in the first day of file: code 43.128, level type 1, level 0. Dismissing task mrsow in table Eday
In order to solve this problem ece2cmor3 has to be modified to read constant field 43.128 only once from ICMGG+000000 file.
@treerink @goord

**Comment by treerink** [comment]
@francocatalano Just checking: did you follow [this wiki recommendation](https://github.com/EC-Earth/ece2cmor3/wiki/Recommended-strategies#cmorisation-of-a-restarted-experiment) for cmorising?

**Comment by francocatalano** [comment]
@treerink In my case I did not start from restart but used IFS initial conditions files. Actually, I used the IC files you generated (it was the beginning of november 2019) for the r1 historical run for 1980-01-01. Therefore, I am starting to cmorize at leg 001 of the simulation.

**Comment by goord** [comment]
Hi @francocatalano I received your files, trying to fix this issue first

**Comment by goord** [comment]
Hi @francocatalano I created a fix in the branch fx-from-ini. Can you test by running the code at least on leg 1 and 2, and processing all variables?

BTW when you check out the branch, don't forget to run python setup.py install again.

**Comment by goord** [comment]
Hi Franco, to test my fix, run in your ece2cmor3 repo
```
git fetch
git checkout  fx-from-ini
conda activate ece2cmor3
python setup.py install
```
and then run your test

**Comment by francocatalano** [comment]
Hi @goord . It worked fine for leg 001. But when I launch it to leg 002 it gets stuck after a while and does not finish even after increasing the maximum walltime to 1h. That is very strange since leg 001 completes in about 20 minutes.

**Comment by goord** [comment]
Hi Franco can you make your leg 002 readable to me? thx

**Comment by francocatalano** [comment]
> Hi Franco can you make your leg 002 readable to me? thx

Hi @goord .Done. Let me know if you have problems accessing the files.

**Comment by goord** [comment]
Hi Franco I noticed you have ```ICM{GG,SH}ECE3+198012``` in your leg 2 directory containing copies of fields at 1981-1-1 00:00:00. This confuses ece2cmor3, it will think that this is the actual first file of the leg. You can remove those files and the cmorization should then work properly.

**Comment by francocatalano** [comment]
Thanks @goord After removing the two ICM{GG,SH}ECE3+198012 files from 002 folder the job still gets stuck and, additionally, I now get this error in the .cmor.log file:
Error: approximate time axis interval is defined as 86400.000000 seconds (1.000000 days), for value 1 we got a difference of 15854400.000000 seconds (183.500000 days), which is 18250.000000 % , seems too big, check your values

**Comment by goord** [comment]
Hi Franco, can you send me that log file so I can check which variable is causing that error?

**Comment by francocatalano** [comment]
@goord here is the file:
[ECE3-ifs-002-20200323091745.cmor.log](https://github.com/EC-Earth/ece2cmor3/files/4370225/ECE3-ifs-002-20200323091745.cmor.log)

**Comment by goord** [comment]
Hi Franco, could you also send me the standard output stream with all messages from ece2cmor3 itself, probably your job output file?

**Comment by francocatalano** [comment]
here they are:
[ECE3-ifs-002-20200323091745.log](https://github.com/EC-Earth/ece2cmor3/files/4370297/ECE3-ifs-002-20200323091745.log)

[pbs-log-for-cmorising-ECE3-ifs-002.out.txt](https://github.com/EC-Earth/ece2cmor3/files/4370318/pbs-log-for-cmorising-ECE3-ifs-002.out.txt)

**Comment by goord** [comment]
Hi Franco I can't reproduce your problem, I ran with the branch at ECMWF your leg 002, and it produced 98 variables in a 14 min. job with 18 threads,

[proc_ifs_ls3mip.txt](https://github.com/EC-Earth/ece2cmor3/files/4382087/proc_ifs_ls3mip.txt)

**Comment by goord** [comment]
...are you sure you have cleaned your tmp directory too? After a crashing or hanging job this may cause problems

**Comment by francocatalano** [comment]
Hi @goord Tried again after removing temp-cmor-dir but got the same error in .cmor.log file.

**Comment by francocatalano** [comment]
here is the job script if it may help dignose the problem
[submit-at-cca-ece2cmor-leg-job-pdLC-ssp126.sh.txt](https://github.com/EC-Earth/ece2cmor3/files/4382274/submit-at-cca-ece2cmor-leg-job-pdLC-ssp126.sh.txt)

**Comment by goord** [comment]
Hi Franco, could you launch one more run with only 1 thread (so change the npp option for ece2cmor3 from 18 to 1 and the EC_threads_per_task = 1 as well). This will make the log more understandable.

**Comment by francocatalano** [comment]
Hi @goord I now managed to get rid of the error. After a few more tests (with 1 thread as well as with 18) I realized that the error arise only if I activate activateece2cmor3 manually before launching the submit script. If I did not manually activate ece2cmor3 it works, otherwise it gives that error. Perhaps, this is due to the fact that ece2cmor3 is already being activated inside the submit script so, manually activating it before somehow creates problems. Thank you for your support!

**Comment by treerink** [comment]
> I realized that the error arise only if I activate activateece2cmor3 manually before launching the submit script. If I did not manually activate ece2cmor3 it works, otherwise it gives that error. Perhaps, this is due to the fact that ece2cmor3 is already being activated inside the submit script so, manually activating it before somehow creates problems. Thank you for your support!

That's weird, I often call the submit script both from the activated environment and from a plane login, and never noticed such differences. Though I am not sure I called the submit script on cca from the activated environment.

**Comment by treerink** [comment]
Ok great Franco that your tests are successful now. I will test once more the `fx-from-ini` branch for the general test-all case, before merging this branch into the master.

**Comment by goord** [comment]
That is quite strange. I thought it shouldn't matter since the script is being executed on a different machine with a clean environment. On the other hand, we don't know exactly how the conda activate script works, and how it may interfere with the cca modules...

Maybe something to add to the documentation

**Comment by treerink** [comment]
Okay will do the general tests with the latest fix. Some delay, I had to regain HPC access.

**Comment by francocatalano** [comment]
quick update. I have launched all the 121 years of one of our LS3MIP experiments. When checking the results with files-per-year.sh I found out that some (about 20 over 121) jobs did produce only 97 files instead of the 98 expected. I have checked the log files and found that some error occurred in reading the data. The good is that when I launched again the jobs with this problem they complete producing all the 98 files as expected. Now, the random errors may be due to some cca problems (yesterday the server underwent some maintenance) or to unpredictable ece2cmor3 behaviour. Just wanted to let you know.

**Comment by treerink** [comment]
Ok yes sounds like glitches, I encountered such glitches as well during cmorising the r1 results, which were indeed solved by rerunning. So yes the checks on the cmorised data are important.

**Comment by francocatalano** [comment]
Just checked the results with nctime and got some overlaps and broken time series:
Number of dataset(s) with overlap(s): 53
Number of dataset(s) with broken time series: 4
I attach the nctime logfile for reference.
[nctcck-20200402-102615-16155.log](https://github.com/EC-Earth/ece2cmor3/files/4420613/nctcck-20200402-102615-16155.log)

**Comment by goord** [comment]
Looks like 1980 has a problem, it should contain the time point 1981-1-1...

**Comment by treerink** [comment]
I just see the message is everywhere totally identical:
```
<-- overlap from 19810101000000 to 19810101000000
```
where the overlap dates are the same (so no real overlap?). 

For the end message:
```
Number of dataset(s) with broken time series: 4
```
Could not spot any messages in the rest of the log, did you spot them?

**Comment by treerink** [comment]
I finished the general test and compared the ERROR logs with the latest ones posted in #542. Though a few small changes which I can't fully trace back, overall it seems fine. So I actually want to merge the `fx-from-ini` branch into the master.

**Comment by francocatalano** [comment]
> I just see the message is everywhere totally identical:
> 
> ```
> <-- overlap from 19810101000000 to 19810101000000
> ```
> 
> where the overlap dates are the same (so no real overlap?).
> 
> For the end message:
> 
> ```
> Number of dataset(s) with broken time series: 4
> ```
> 
> Could not spot any messages in the rest of the log, did you spot them?
look for "BREAK" in the log file.
lines: 980, 1063, 1887, 3057, 3095, 3106, and so on...

**Comment by francocatalano** [comment]
Looking a bit more at the files I found another strange thing: most of the files span yyyy0101-yyyy1231 while a few of them span a shorter period:
mrsow_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_19800101-19810101.nc
mrsow_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_20620101-20620126.nc
just to mention some of them. Any thoughts about possible reasons?

**Comment by goord** [comment]
Hi Franco, it could be still an issue with ece2cmor3, especially since it is mrsow causing these issues again. To figure out the problem I would like to have a look at the intermediate grib and nc file, but those are probably cleaned already.

What happens if run the leg once more?

**Comment by francocatalano** [comment]
@goord To sum up, we currently have two problems:
1) broken time series reported by nctime. This error does occur on the following files:
mrsow_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_19800101-19810101.nc
snc_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_198001-198012.nc
hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_198001-198101.nc
hfdsn_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_19800101-19810101.nc

2) files produced with weird date intervals, as reported above. The files presenting this problem are:
mrsow_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_19800101-19810101.nc
mrsow_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_20620101-20620126.nc
snc_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_203101-203102.nc
hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_198001-198101.nc
hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_201701-201709.nc
hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_202701-202707.nc
hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_203101-203102.nc
hfdsn_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_19800101-19810101.nc
hfdsn_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_19980101-19980122.nc
hfdsn_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_20100101-20100519.nc
hfdsn_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_20620101-20620126.nc
hfdsn_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_20650101-20650404.nc
hfdsn_Eday_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_20990101-20990312.nc

Then, there are the strange messages about the overlaps which can be most probably ignored.

**Comment by francocatalano** [comment]
I am now re-running all the legs (as it was not clear at all which legs had problems). I'll keep you updated.

**Comment by treerink** [comment]
Just to make sure: take care the EC-Earth output directories are clean: so only files from your latest run are present and not files from older runs (though the recent `ece2cmor` should give warnings for this). And secondly take care when repeating your cmorising jobs again, that they end up in a brant new empty directory. This just to avoid painful digging around. Probably you did this all correct, but better to stress this once more.

**Comment by francocatalano** [comment]
All the legs complete. Two of them have this error:
cdo(3) expr: Process started
cdo(4) selcode: Process started
Error (grb_read_record): Failed to read GRIB record

I confess that I am a bit scared about these random errors. Two days ago re-running the legs solved 
them but then I discovered the other problems. There should be a bug somewhere.

I attach the relative logfiles.
[ECE3-ifs-070-20200402192136.cmor.log](https://github.com/EC-Earth/ece2cmor3/files/4425794/ECE3-ifs-070-20200402192136.cmor.log)
[ECE3-ifs-070-20200402192136.log](https://github.com/EC-Earth/ece2cmor3/files/4425795/ECE3-ifs-070-20200402192136.log)
[ECE3-ifs-105-20200402195937.cmor.log](https://github.com/EC-Earth/ece2cmor3/files/4425796/ECE3-ifs-105-20200402195937.cmor.log)
[ECE3-ifs-105-20200402195937.log](https://github.com/EC-Earth/ece2cmor3/files/4425797/ECE3-ifs-105-20200402195937.log)

**Comment by francocatalano** [comment]
> Just to make sure: take care the EC-Earth output directories are clean: so only files from your latest run are present and not files from older runs (though the recent `ece2cmor` should give warnings for this). And secondly take care when repeating your cmorising jobs again, that they end up in a brant new empty directory. This just to avoid painful digging around. Probably you did this all correct, but better to stress this once more.

@treerink Yes, I did make sure all the above. Thanks

**Comment by francocatalano** [comment]
I have re-run the legs with error and, as expected, they completed without any error. Then I checked the results with nctime and, again, got a couple of datasets with broken timeseries. I attach the nctime logfile
[nctcck-20200406-161340-32138.log](https://github.com/EC-Earth/ece2cmor3/files/4439660/nctcck-20200406-161340-32138.log)

As you can see (look for BREAK in the file), the broken time series error involves many files (all those with wrong date intervals in the name) and many legs. Furthermore, these errors did not pop up in the ece2cmor3 log files.
Having re-run ece2cmor3 over different days and got the same kind of random errors I would exclude the errors are caused by cca maintenance (the server has not been in maintenance for so many days).

**Comment by goord** [comment]
Hi Franco I think you are right, and I bet that you do not get these errors if you run with the 1.4 version (and leave out mrsow). In other words, I believe my changes to be able to cmorize mrsow may have caused a bug in ece2cmor3... 

I observe that the BREAK statements always occur in hfsdn_Eday_* files, which happens to be the only variable being cmorized in that table other than... mrsow (BTW you're sure you have updated your branch and re-installed ece2cmor3 right?)

I believe the bug you see happens if ece2cmor3 uses mrsow to create that time axis in that table (this choice is random) and something went wrong in either filtering or post-processing. Can you send me ece2cmor3 log output for one of those incorrect legs to test this hypothesis?

Then there is the first leg 1980. Not sure what's going on there, it should not include the time point 1-1-1981 0:0:0. I will have a look at it on cca (you have given me permissions rigt?).

**Comment by goord** [comment]
By the way I just discovered looking at these variables that the radiative fluxes are missing in hfdsl... see #620

**Comment by goord** [comment]
By the way @francocatalano the previous run you did was with the latest version of the branch right? And with 18 threads?

**Comment by goord** [comment]
@treerink understood, will work on it today

**Comment by goord** [comment]
Hi Franco, I am able to recreate your problem, the issue is with the filtering of the heat flux in snow, I somehow must have introduced a bug there

**Comment by francocatalano** [comment]
Hi @goord Thenks. Let me know if you need more info from my side.

**Comment by goord** [comment]
Hi @francocatalano it turns out my changes to fix the mrow issue have prevented the shift of accumulated variables during grib filtering, which makes your cmorized data incorrect.

I have committed a fix on the master branch and I am currently testing it, you can do the same.

**Comment by francocatalano** [comment]
Ok @goord  How do I get the fix?

**Comment by treerink** [comment]
@francocatalano :
```
git checkout master
git pull
python setup.py develop   # in your ece2cmor3 root directory
```

**Comment by goord** [comment]
I still get weird time bounds for hfdsn in the LImon table every now and then, it's quite strange. The grib file looks fine but the post-processed nc file often way too short...

**Comment by goord** [comment]
If I then process the filtered grib file with exactly the same cdo command, I get a correct netcdf file which makes the bug hard to pinpoint

**Comment by francocatalano** [comment]
what is very strange (and worrisome!) is that the code is not displaying a deterministic behavior. Could it be that the cdo command called from python uses, for some reason, a different version of cdo?

**Comment by goord** [comment]
Intermediate findings:

* The grib files leading to ```hfdsn_LImon_*.nc``` all seem to be correct, and running cdo manually on the cca head node on these flies produces correct nc files. 
* ece2cmor3 in single-threaded mode produces the correct files.

I am now running ece2cmor3 with less threads on cca

**Comment by goord** [comment]
Hi @francocatalano when I run with ```--npp 10```, but I allocate 18 threads on cca like you I don't encounter any errors. It could be that if you run with the maximal 18 threads cdo somehow goes out of memory (silently apparently) during the processing of hfdsn_LImon.nc. I suggest you either run ece2cmor3 on a parallel queue node with 18 threads or in the fractional queue with 10 threads or less. In the latter case make sure to allocate still 18 cores.

The issue is a bit disturbing as we get no messages whatsoever from cdo it failed to process a variable...

**Comment by francocatalano** [comment]
Hi @goord Thanks. I'll give it a try. I can get all the latest fixes by checking out master, right?

**Comment by goord** [comment]
Yes

**Comment by francocatalano** [comment]
Hi @goord I've just tried (after having switched to master branch of ece2cmor3) your suggestion of setting npp to 10. I have launched 50 legs at once. Four of them give the following error:

2020-04-08 09:18:17 ERROR:ece2cmor3.cdoapi: (returncode:1) cdo(2) monmean: Process started
cdo(3) expr: Process started
cdo(4) selcode: Process started
Warning (cgribexScanTimestep1): Record  5 (id=141.128 lev1=0 lev2=0) timestep 1: Inconsistent verification time!
Error (grb_read_record): Failed to read GRIB record

I don't think testing a further reduction of npp would be the solution. Most probably it would work, by chance, on a number of legs and still give errors on few of them.

**Comment by treerink** [comment]
@goord here what you asked for, they are results of several of recent tests:
```
find . -name hfdsn_LImon_*.nc|sort

./cmor-cmip-test-all-1/t001/ifs/001/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200318/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199001-199012.nc
./cmor-cmip-test-all-2/t002/ifs/001/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200320/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199001-199012.nc
./cmor-cmip-test-all-3/t002/ifs/001/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199001-199012.nc
./cmor-cmip-test-all-3/t002/ifs/002/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199101-199112.nc
./cmor-cmip-test-all-3/t002/ifs/003/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199201-199212.nc
./cmor-cmip-test-all-3/t002/ifs/004/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199301-199312.nc
./cmor-cmip-test-all-3/t002/ifs/005/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199401-199412.nc
./cmor-cmip-test-all-3/t002/ifs/006/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199501-199512.nc
./cmor-cmip-test-all-3/t002/ifs/007/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199601-199612.nc
./cmor-cmip-test-all-3/t002/ifs/008/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199701-199712.nc
./cmor-cmip-test-all-3/t002/ifs/009/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199801-199812.nc
./cmor-cmip-test-all-3/t002/ifs/010/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199901-199912.nc
./cmor-cmip-test-all-4/t002/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199001-199012.nc
./cmor-cmip-test-all-4/t002/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199101-199112.nc
./cmor-cmip-test-all-4/t002/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199201-199212.nc
./cmor-cmip-test-all-4/t002/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199301-199312.nc
./cmor-cmip-test-all-4/t002/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199501-199512.nc
./cmor-cmip-test-all-4/t002/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199701-199712.nc
./cmor-cmip-test-all-4/t002/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199801-199812.nc
./cmor-cmip-test-all-4/t002/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200326/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199901-199912.nc
./cmor-cmip-test-all-5/t002/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3/piControl/r1i1p1f1/LImon/hfdsn/gr/v20200402/hfdsn_LImon_EC-Earth3_piControl_r1i1p1f1_gr_199001-199101.nc
```
It looks like that only in the most recent one december is not reached?

**Comment by goord** [comment]
Hi @francocatalano , do you still have the temporary files belonging to those failed legs? I want to have a look at the filtered grib files containing snow depth (141). Can you verify that for the failed legs hfdsn_LImon_*.nc is missing?

Sorry to bother you like that, this is really a nasty bug popping up once in a while. If you want to proceed you probably can re-cmorize the failed 5 legs and be done with it, but from my perspective it is essential to understand what's going on

**Comment by francocatalano** [comment]
Hi @goord you can find all the temp files here:
/scratch/ms/it/ccfc/cmorisation/temp-cmor-dir
The failed legs are: 014, 017, 024, 031
I have opened read permissions but let me know in case of problems.

I don't find very useful for me to proceed now launching again the failed jobs. There is still the possibility that other errors will pop up when checking with nctime. Please let me know as soon as the bug is completely fixed. And, of course, if I can help with more info. Thanks a lot for your effort.

**Comment by francocatalano** [comment]
> Can you verify that for the failed legs hfdsn_LImon_*.nc is missing?

@goord I confirm. hfdsn is missing for all the failed legs.

**Comment by goord** [comment]
Hi Franco, can you make your experiment output readable for me (now I can only access leg 1 & 2) so I can launch a large number of simultaneous (yet distinct) jobs myself? thx

**Comment by francocatalano** [comment]
Hi @goord Done.

**Comment by goord** [comment]
Franco I notice the output is missing 0-hour fields on january 1st for every year, same problem as #612

**Comment by francocatalano** [comment]
> Hi Franco I noticed you have `ICM{GG,SH}ECE3+198012` in your leg 2 directory containing copies of fields at 1981-1-1 00:00:00. This confuses ece2cmor3, it will think that this is the actual first file of the leg. You can remove those files and the cmorization should then work properly.

Hi @goord You recommended removing these files. I can copy again them from ECFS if needed. Let me know.

**Comment by goord** [comment]
Ah I thought those were copied timestamps... Could you append each of these to the december gridpoint file of the previous year?

**Comment by francocatalano** [comment]
@goord I am not sure I understood what you asked for. What I can do is to copy the ICM{GG,SH}ECE3+{yyyy-1}12 in the output leg folders.
The timestamp of these files is yyyy-01-01 00:00:00
I have copied back the 002 files. Could you please check and confirm these are the files needed. Thanks.

**Comment by goord** [comment]
Hi @francocatalano the problem is that ece2cmor3 may get confused with the current setup because it will find the ICMGGECE3+yyyy12 files twice. I have programmed ece2cmor3 such that it will look for those files anywhere in the output/ifs directory (it moves one directory level up) to get the zero-hour 01/01 fields and this search will give back 2 results now.

One option would be to run ece2cmor3 in a symlink of the leg with an extra level, e.g.
```
work/002/output -> ECE3/output/ifs/002
```

Another option would be to append the 'previous' december grib files for leg x to the actual december grib files of leg x-1

**Comment by francocatalano** [comment]
Hi @goord. I think that manipulating the grib files can be difficult (cdo cannot work with ICMGG files) and dangerous too since there is the risk of introducing changes in the file structure that may lead to additional problems. Therefore, I would go with the other alternative. May it help if I restore the ICM{GG,SH}ECE3+{yyyy-1}12 in a subfolder, let's say ECE3/output/ifs/002/prev_year ?

**Comment by goord** [comment]
Hi @francocatalano it's the other way around: the December file of the previous leg has to become invisible because it contains no useful information. Wait, I'll create a branch to let ece2cmor3 handle this situation automatically...

**Comment by francocatalano** [comment]
> Hi @francocatalano it's the other way around: the December file of the previous leg has to become invisible because it contains no useful information. Wait, I'll create a branch to let ece2cmor3 handle this situation automatically...

ok. Thanks.

**Comment by goord** [comment]
Hi Franco, fixing this in ece2cmor3 appears more complicated than I anticipated but the trick with the links should work if you restore all those december files from ecfs, you just include
```bash
printf -v CURLEG "%03d" $LEG
printf -v PRVLEG "%03d" $((LEG-1))
WORKDIR=$SCRATCH/ls3mip/work/$PBS_JOBID
mkdir -p $WORKDIR/$CURLEG
mkdir -p $WORKDIR/$PRVLEG
SRCDIR=/scratch/ms/it/ccfc/ECEARTH-RUNS_ls3mip_pdLC_ssp126/ECE3/output/ifs/${LEG}
for f in $SRCDIR/ICM{GG,SH}ECE3+$((1980+LEG-1))?? ; do
    ln -s $f $WORKDIR/$CURLEG/$(basename $f)
done
for f in $SRCDIR/ICM{GG,SH}ECE3+$((1980+LEG-2))12 ; do
    ln -s $f $WORKDIR/$PRVLEG/$(basename $f)
done
for f in /scratch/ms/it/ccfc/ECEARTH-RUNS_ls3mip_pdLC_ssp126/ECE3/output/ifs/001/ICM{GG,SH}ECE3+000000 ; do
    ln -s $f $WORKDIR/$(basename $f)
done
```
in the submission and let ```ece2cmor3``` run on ```$WORKDIR/$CURLEG``` instead of the original directory

**Comment by francocatalano** [comment]
> ```shell
> for f in $SRCDIR/ICM{GG,SH}ECE3+$((1980+LEG-1))?? ; do
>     ln -s $f $WORKDIR/$CURLEG/$(basename $f)
> done
> ```

@goord ECE3+$((1980+LEG-2))12 does not exist for LEG 001. ece2cmor3 should take into account this

**Comment by francocatalano** [comment]
> ```shell
> for f in $SRCDIR/ICM{GG,SH}ECE3+$((1980+LEG-1))?? ; do
>     ln -s $f $WORKDIR/$CURLEG/$(basename $f)
> done
> for f in $SRCDIR/ICM{GG,SH}ECE3+$((1980+LEG-2))12 ; do
>     ln -s $f $WORKDIR/$PRVLEG/$(basename $f)
> done
> ```

further: with the above lines of code the job for LEG+1 tries to overwrite the symlinks created by the job for LEG. The problem is, considering LEGs 003 and 004:
ICMGGECE3+198212 in /scratch/ms/it/ccfc/ECEARTH-RUNS_ls3mip_pdLC_ssp126/ECE3/output/ifs/004
has the same name of ICMGGECE3+198212 in /scratch/ms/it/ccfc/ECEARTH-RUNS_ls3mip_pdLC_ssp126/ECE3/output/ifs/003
but they are NOT the same file. Actually, the former contain only one time step while the latter contain all december timesteps.
Therefore, the above cannot work

**Comment by francocatalano** [comment]
ah, ok. I have to put that inside the pbs job in order that PBS_JOBID is defined

**Comment by francocatalano** [comment]
@goord I've tried to do that but still $PBS_JOBID is not defined. I attach the submission script. Please have a look at it and let me know. Thanks.
[submit-at-cca-ece2cmor-leg-job-pdLC-ssp126_fix.sh.txt](https://github.com/EC-Earth/ece2cmor3/files/4457851/submit-at-cca-ece2cmor-leg-job-pdLC-ssp126_fix.sh.txt)

**Comment by goord** [comment]
@francocatalano I didn' realize you were using Thomas' job script generator. The job id is just there to create a unique directory per job, but you can use the leg for that too, e.g.
```
WORKDIR='$running_directory/temp_work/ifs_$CURLEG'
```
Then the script should create directories
```
temp_work
    |_ifs_002
        |_001
        |_002
    |_ifs_003
        |_002
        |_003
    |...
```
in which ece2cmor3 runs and can only see the correct symlinks

**Comment by francocatalano** [comment]
@goord The latest fixes are still on the master branch, right?

**Comment by goord** [comment]
Right

**Comment by francocatalano** [comment]
Hi @goord I am still getting random errors of the kind:
2020-04-10 09:46:13 ERROR:ece2cmor3.cdoapi: (returncode:1) cdo(2) daymean: Process started
cdo(3) expr: Process started
cdo(4) selcode: Process started
Error (grb_read_record): Failed to read GRIB record

and:

2020-04-10 10:03:47 ERROR:ece2cmor3.cdoapi: (returncode:1) cdo(2) monmean: Process started
cdo(3) expr: Process started
cdo(4) selcode: Process started
Error (grb_read_record): Failed to read GRIB record

Looking at the files, hfdsn_*.nc is still missing for Eday (first kind of error) or LImon (second kind of error)
Here is the submit script:
[submit-at-cca-ece2cmor-leg-job-pdLC-ssp126_fix.sh.txt](https://github.com/EC-Earth/ece2cmor3/files/4461152/submit-at-cca-ece2cmor-leg-job-pdLC-ssp126_fix.sh.txt)

and here the logfiles:
[ECE3-ifs_010-010-20200410093728.cmor.log](https://github.com/EC-Earth/ece2cmor3/files/4461154/ECE3-ifs_010-010-20200410093728.cmor.log)
[ECE3-ifs_010-010-20200410093728.log](https://github.com/EC-Earth/ece2cmor3/files/4461155/ECE3-ifs_010-010-20200410093728.log)
[ECE3-ifs_040-040-20200410095332.cmor.log](https://github.com/EC-Earth/ece2cmor3/files/4461156/ECE3-ifs_040-040-20200410095332.cmor.log)
[ECE3-ifs_040-040-20200410095332.log](https://github.com/EC-Earth/ece2cmor3/files/4461157/ECE3-ifs_040-040-20200410095332.log)
[ECE3-ifs_043-043-20200410100439.cmor.log](https://github.com/EC-Earth/ece2cmor3/files/4461158/ECE3-ifs_043-043-20200410100439.cmor.log)
[ECE3-ifs_043-043-20200410100439.log](https://github.com/EC-Earth/ece2cmor3/files/4461159/ECE3-ifs_043-043-20200410100439.log)


Additionally, I have noticed that the logfiles have been renamed and now include twice the leg number, which prevents them to be correctly moved to cmorised-results/cmor-ls3mip-cmip/ECE3/logs/.

**Comment by francocatalano** [comment]
wait...I've just realized that did not update the nested cmor tables. Running the test again.

**Comment by francocatalano** [comment]
@goord Unfortunately, the test with the updated nested cmor tables is giving the same kind of errors.

**Comment by goord** [comment]
Hi @francocatalano could you change the permissions to the restored ```ICM{SH,GG}ECE3+(yyyy-1)12``` files as well? Thanks

**Comment by francocatalano** [comment]
@goord Ops...I overlooked that. Now it's done.

**Comment by goord** [comment]
Hi Franco, I figured out you do have to concatenate the december files to get the correct results with ece2cmor3, because now the fluxes that been accumulated in the last 3 hours of december will be in the next leg and therefore not noticed by ece2cmor3.

I advise you to move all december files to a backup location and the single time-slice files to somewhere else, and then use the linux cat:
```
cat backup/ICMGGECE3+198012 backup2/ICMGGECE3+198012 > 001/ICMGGECE3+198012
cat backup/ICMSHECE3+198012 backup2/ICMSHECE3+198012 > 001/ICMSHECE3+198012
``` 
to produce the normal EC-Earth output form. Then update ece2cmor3 to the latest master and test the cmorization. I have tested on cca and got still some errors in hfdsn_LImon_* with too few time stamps, but at least the rest of your data will be ok.

**Comment by francocatalano** [comment]
Hi @goord I have updated ece2cmor3 and the nested cmor tables and done a test following your latest suggestion. Leg 001 was processed correctly but I got the following errors for the other legs:
2020-04-14 17:35:45 ERROR:ece2cmor3.grib_filter: Assumed previous month file for /scratch/ms/it/ccfc/cmorisation//temp_work/ifs_002/002/ICMSHECE3+198101: /scratch/ms/it/ccfc/cmorisation/temp_work/ifs_002/ICMSHECE3+000000, this is probably not correct!

and:
2020-04-14 17:44:45 ERROR:ece2cmor3.ifs2cmor: CMOR failed to load table Amon, the following variable will be skipped: hfls. Reason: Problem with 'cmor.load_table'. Please check the logfile (if defined).

and in the .cmor.log files:
! Error: approximate time axis interval is defined as 86400.000000 seconds (1.000000 days), for value 1 we got a difference of 15854400.000000 seconds (183.500000 days), which is 18250.000000 % , seems too big, check your values

These are the logfiles for leg 002:
[ECE3-ifs_002-002-20200414173544.cmor.log](https://github.com/EC-Earth/ece2cmor3/files/4476925/ECE3-ifs_002-002-20200414173544.cmor.log)
[ECE3-ifs_002-002-20200414173544.log](https://github.com/EC-Earth/ece2cmor3/files/4476926/ECE3-ifs_002-002-20200414173544.log)

This is becoming very frustrating.
I attach my submit script:
[submit-at-cca-ece2cmor-leg-job-pdLC-ssp126_fix.sh.txt](https://github.com/EC-Earth/ece2cmor3/files/4476931/submit-at-cca-ece2cmor-leg-job-pdLC-ssp126_fix.sh.txt)

Please, revise it and let me know whether I misinterpreted somehow what you suggested. Thanks.

**Comment by goord** [comment]
Hi Franco, sorry you can remove the ``ifs_$CURLEG`` inner subdirectory from your WORKDIR, in this way you mimic the normal output structure of ifs such that ece2cmor3 can find the correct December files in the previous leg directory.

BTW you can create all those concatenated files and links once.

**Comment by treerink** [comment]
Hi, I want to close this issue, the `mrsow` issue is solved. For follow up questions or assistance for handling the data during cmorisation please open a new issue. Also further discussion of `hfdsn` `LImon` should be continued in a separate issue if necessary. Currently it looks like we have to accept that at cca we sometimes have a glitch with `hfdsn` `LImon` and that just rerunning `ece2cmor3` will fix it.

**Comment by francocatalano** [comment]
Hi @goord  I have done that. Now, apparently, there are no errors reported in the cmorization logfiles. Nevertheless, some of the legs still randomly produce incomplete time series for the variable hfdsn_LImon_*.
@treerink Before closing the issue I would like at least to re-run nctime and see.

**Comment by goord** [comment]
Hi Franco, I noticed that as well and, to be honest, I don't know what's causing this. It may be something with memory in the fractional queue of cca, or something else. Tomorrow I will do a test in the parallel queue.

**Comment by francocatalano** [comment]
Hi @goord Please also note that while for some legs the incomplete time series is evidenced in the file name, for others it is not. That is, the file name is correct but the dimension is much smaller than the other files for the same variable. For example:
-rw-r----- 1 ccfc it 1,4M 15 apr 16.55 hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_207801-207812.nc
-rw-r----- 1 ccfc it 1,3M 15 apr 16.59 hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_207901-207912.nc
-rw-r----- 1 ccfc it 823K 15 apr 16.58 hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_208001-208007.nc
-rw-r----- 1 ccfc it 450K 15 apr 16.59 hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_208101-208112.nc
-rw-r----- 1 ccfc it 1,3M 15 apr 17.03 hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_208201-208212.nc
-rw-r----- 1 ccfc it 1,1M 15 apr 17.02 hfdsn_LImon_EC-Earth3_amip-lfmip-pdLC_r1i1p1f1_gr_208301-208310.nc

which makes very painful to assess whether ece2cmor3 produced the expected results or not.

**Comment by goord** [comment]
Hi Franco, I tried the parallel queue, without succes. I also noticed that the filtered grib files corresponding to hfdsn, the file named 147.128.1_146.128.1_177.128.1_176.128.1.3_141.128.1.3, is never cleaned up... this may give us a lead to the problem

**Comment by goord** [comment]
Hi Franco it seems that somehow the tool is mixing up hfdsl and hfdsn once in a while...

**Comment by francocatalano** [comment]
@goord Could it be a problem of misinterpreted wildcards in the code, somehow? Just guessing...

**Comment by goord** [comment]
Hi Franco, the problem is wirt simultaneous threads writing to the same file (I believe). Going to fix it asap.

**Comment by goord** [comment]
Hi @francocatalano I created the branch flux-inst-bugfix (the issue originates from the fact that hfdsn is a combination of fluxes and instantaneous snow depth) which may solve your problem. I ran 25 years on your data and no longer saw the missing time stamps in the hfdsn files.

To test, run ```git checkout flux-inst-bugfix``` in your ece2cmor3 repo, and then in your conda ece2cmor3 environment reinstall (i.e. ```python setup.py install```), and resubmit your job.

@treerink it would be nice if the branch, if validated, could end up in the release

**Comment by goord** [comment]
BTW Franco, note that you still have to run in the directory with the concatenated december files (or create them in the submission script).

**Comment by treerink** [comment]
Ok, @francocatalano if you ley me know whether the glitch doen not longer occur with the `flux-inst-bugfix` branch, then I will merge the branch and thereafter do a final `test-all` test before releasing #605.

**Comment by francocatalano** [comment]
@goord just tried to get the fix but got the following error:
git checkout flux-inst-bugfix
error: pathspec 'flux-inst-bugfix' did not match any file(s) known to git
what's wrong?

**Comment by treerink** [comment]
Hi Franco, first pull in your master (or anywhere):
``` 
git pull
git checkout flux-inst-bugfix
python setup.py install          # in your ece2cmor3 root dir
```

**Comment by francocatalano** [comment]
Ops...I forgot pull. Thanks @treerink Running the test now. Then, I will also check results with nctime. I'll let you know as soon as the test and the checks are done.

**Comment by goord** [comment]
Hi Franco, sorry about that, I was a bit enthousiastic in refactoring some code in that branch, could you update (pull) and try once more?

**Comment by francocatalano** [comment]
Hi @goord @treerink The cmorization with the new flux-inst-bugfix branch completed without any error in the logfiles. The timeseries of hfdsl and hfdsn look complete (both dimension and time stamp of the files are reasonable). I have run nctime and it completed without any error as well:

:: COMMAND :: /home/ms/it/ccfc/.local/bin/nctcck -i esgini-dir -l log-nctcck CMIP6
Number of node(s): 11858
Number of dataset(s): 98
Number of dataset(s) with overlap(s): 0
Number of dataset(s) with broken time series: 0
Number of file(s) scanned: 11858
Number of file(s) skipped: 0

@treerink you can proceed with test-all. After that, if your tests are ok, I'll close the issue.
Thank you guys for the very good job!

**Comment by treerink** [comment]
The #624 is merged into the master, fixing the remaining `hfdsl` issue in this issue.
